### DISCUSSION TITLE
Issues migrating 4.15.2 database

### DISCUSSION THREAD

**Discussion by dalers** [timschofield/webERP#210] [Problems / Bugs] [answered]
I'm trying to restore a vanilla webERP database dumped 2018-10-01 from what I assume was a v4.15.2 system. After accepting login credentials the new page doesn't render, and then when reloaded the Database Upgrade script runs. Is this expected? This reply from Tim implies it should not (https://github.com/timschofield/webERP/discussions/204#discussioncomment-11061609), but also that there should be a DBUpgradeNumber confname record in the conf table, which there does not seem to be. Is the DBUpgradeNumber variable new since 4.15.2? Should migrating an old database work?

Also, fwiw, the current installer creates 178 tables but the restored database only has 157 tables. Is this expected (and correct)?

Login to weberp using admin user and password in recovered database

![2024-11-07 15_09_15-WebERP Login screen](https://github.com/user-attachments/assets/5cfb542a-a9b0-40ee-b95b-8a1faee3d514)

Blank page after clicking Enter

![2024-11-07 15_09_48-weberp dalescott net](https://github.com/user-attachments/assets/131890dd-48fb-48b3-96ee-3aaf96b42f48)

Reloaded page and now UpdateDatabase script is running

![2024-11-07 15_14_38-Database Upgrade](https://github.com/user-attachments/assets/e7ce0bbd-2e77-47e4-aa83-35f24a8ec0e4)

Examining conf table, confname DBUpgradeNumber does not seem to exist.

![2024-11-07 15_20_46 no confname DBUpdateNumber](https://github.com/user-attachments/assets/0fcd37dd-e71c-466e-9102-6626223ad4b6)

I'll post the .sql.gz file if it would help (80KB)

Cheers,
Dale

**Comment by dalers** [comment]
Found a couple errors in /var/log/httpd-error.log after entering login credentials ("This page isn't working").

`[Thu Nov 07 15:09:40.502081 2024] [php:error] [pid 33071] [client 162.223.103.50:25749] PHP Fatal error:  Uncaught mysqli_sql_exception: Unknown column 'commissionsact' in 'field list' in /usr/local/www/weberp/includes/ConnectDB_mysqli.inc:58\nStack trace:\n#0 /usr/local/www/weberp/includes/ConnectDB_mysqli.inc(58): mysqli_query()\n#1 /usr/local/www/weberp/includes/GetConfig.php(84): DB_query()\n#2 /usr/local/www/weberp/includes/UserLogin.php(169): include('...')\n#3 /usr/local/www/weberp/includes/session.php(171): userLogin()\n#4 /usr/local/www/weberp/index.php(4): include('...')\n#5 {main}\n  thrown in /usr/local/www/weberp/includes/ConnectDB_mysqli.inc on line 58, referer: https://weberp.dalescott.net/`

`[Thu Nov 07 15:10:11.699625 2024] [php:error] [pid 32082] [client 162.223.103.50:64118] PHP Fatal error:  Uncaught mysqli_sql_exception: Table 'weberp.modules' doesn't exist in /usr/local/www/weberp/includes/ConnectDB_mysqli.inc:58\nStack trace:\n#0 /usr/local/www/weberp/includes/ConnectDB_mysqli.inc(58): mysqli_query()\n#1 /usr/local/www/weberp/includes/MainMenuLinksArray.php(13): DB_query()\n#2 /usr/local/www/weberp/index.php(7): include('...')\n#3 {main}\n  thrown in /usr/local/www/weberp/includes/ConnectDB_mysqli.inc on line 58, referer: https://weberp.dalescott.net/`

`[Thu Nov 07 16:48:58.862963 2024] [php:error] [pid 34520] [client 162.223.103.50:51750] PHP Fatal error:  Uncaught mysqli_sql_exception: Unknown column 'commissionsact' in 'field list' in /usr/local/www/weberp/includes/ConnectDB_mysqli.inc:58\nStack trace:\n#0 /usr/local/www/weberp/includes/ConnectDB_mysqli.inc(58): mysqli_query()\n#1 /usr/local/www/weberp/includes/GetConfig.php(84): DB_query()\n#2 /usr/local/www/weberp/includes/UserLogin.php(169): include('...')\n#3 /usr/local/www/weberp/includes/session.php(171): userLogin()\n#4 /usr/local/www/weberp/index.php(4): include('...')\n#5 {main}\n  thrown in /usr/local/www/weberp/includes/ConnectDB_mysqli.inc on line 58, referer: https://weberp.dalescott.net/`

**Comment by dalers** [comment]
Attempted to "Continue with Updates".....

Not clear if I need to click all the Blue arrows, but I did anyway, except for the last two Updates 6 and 7, which do not rotate or show anything additional,

![2024-11-07 17_06_41-Database Upgrade](https://github.com/user-attachments/assets/d498a1ee-aac3-41bf-9a54-a7c8fd6f9698)

and then there seems to be an error after clicking "Continue with Updates"...

![2024-11-07 17_11_26 hmmm, there seems to be an error](https://github.com/user-attachments/assets/884160a3-f758-4fd2-a067-bcd744455266)

(reloading the page shows the same output)

httpd-error.log now includes:

`[Thu Nov 07 17:02:31.067269 2024] [php:error] [pid 34518] [client 162.223.103.50:52093] PHP Fatal error:  Uncaught mysqli_sql_exception: Table 'weberp.modules' doesn't exist in /usr/local/www/weberp/includes/ConnectDB_mysqli.inc:58\nStack trace:\n#0 /usr/local/www/weberp/includes/ConnectDB_mysqli.inc(58): mysqli_query()\n#1 /usr/local/www/weberp/includes/MainMenuLinksArray.php(13): DB_query()\n#2 /usr/local/www/weberp/index.php(7): include('...')\n#3 {main}\n  thrown in /usr/local/www/weberp/includes/ConnectDB_mysqli.inc on line 58, referer: https://weberp.dalescott.net/`

`[Thu Nov 07 17:07:07.778577 2024] [php:error] [pid 34517] [client 162.223.103.50:52166] PHP Fatal error:  Uncaught mysqli_sql_exception: Data too long for column 'reportlink' at row 1 in /usr/local/www/weberp/includes/ConnectDB_mysqli.inc:58\nStack trace:\n#0 /usr/local/www/weberp/includes/ConnectDB_mysqli.inc(58): mysqli_query()\n#1 /usr/local/www/weberp/Z_UpgradeDatabase.php(36): DB_query()\n#2 /usr/local/www/weberp/includes/UpgradeDB_mysqli.php(130): executeSQL()\n#3 /usr/local/www/weberp/sql/updates/5.php(33): NewModule()\n#4 /usr/local/www/weberp/Z_UpgradeDatabase.php(124): include('...')\n#5 {main}\n  thrown in /usr/local/www/weberp/includes/ConnectDB_mysqli.inc on line 58, referer: https://weberp.dalescott.net/Z_UpgradeDatabase.php`

Examining the database directly, the config table now includes DBUpdateNumber with value 4, which I presume proves Update 0 executed, but does it also mean Update 4 completed successfully and Update 5 failed?

![image](https://github.com/user-attachments/assets/09d6de9f-362a-4a7a-821f-d0b03ab54c1f)

This is probably as far as I'll get on my own, what should I do next?

I've attached the database file in case you want to look deeper yourself. 

Cheers,
Dale

[weberp_scc-20181001-weberp.sql.gz](https://github.com/user-attachments/files/17670959/weberp_scc-20181001-weberp.sql.gz)

**Comment by dalers** [comment]
Went a little further....

![2024-11-07 17_35_37 whhhaaat - only updates 5, 6 and 7 now](https://github.com/user-attachments/assets/21dc2857-2e2c-4995-a595-057e958330ad)

Selecting Update 5 ONLY....

![2024-11-07 17_40_21 one update at a time 2](https://github.com/user-attachments/assets/15df6305-67cf-4882-aee1-7cd768325757)

![2024-11-07 17_41_10-Database Upgrade](https://github.com/user-attachments/assets/490abdff-c033-499c-aa89-9db4e13bef88)

![2024-11-07 17_44_06 DBUpdateNumber not updated this time](https://github.com/user-attachments/assets/257c30e6-f336-4a3b-9d1f-af55e0b665e0)

**Comment by timschofield** [comment]
Hi Dale. There were a couple of releases after October 2018. I have archived every release of webERP since v2.9 here - https://github.com/timschofield/webERP/tags
So you would need to run the sql/mysql/upgrade4.15-4.15.1.sql and sql/mysql/upgrade4.15.1-4.15.2.sql scripts before you start. 

With the latest code, the upgrades work differently. I wrote a wiki page on it here - https://github.com/timschofield/webERP/wiki/DBUpgrade

I am thinking I need to go back and check these work with sql-mode set to it's strictest level.

Thanks for all your help,
Tim

**Comment by dalers** [comment]
Hi Tim, I'll start with v4.14 (2018-04-20) comparing the migration scripts to my database, then perform the migration scripts starting from when there's a difference. If you're still awake, what's the easiest/best way to run the migration scripts from the command line?

Thanks for link to doc for current upgrade process.

**Comment by timschofield** [comment]
I prefer to use the command line, but I'm an old fashioned guy!

Phil always preferred phpMyAdmin so it'd whatever suits you most.

**Comment by dalers** [comment]
I found database version in my backup dump was actually 4.15, and I successfully upgraded it to 14.5.2 using the old scripts. I then logged in using the latest code and had the same behavior as before (blank screen after logging in, reloading caused the database upgrade script to run, and upgrading failed). I'll repeat the procedure to confirm behavior and post error messages when I get a chance.

**Comment by timschofield** [comment]
What do the web server error logs say when you try to log in?

Thanks
Tim

**Reply by efeoli** [reply]
Hello Dale and Tim,
I would like to login in if you don’t mind. I am trying to achieve what Dale is doing: rescue my 4.15.2 data into the PHP8 environment of coming webERP v5.0.

When the imported database is first logged in, behavior is as described by Dale. But what happens it is partial upgrade of database adding 10 tables to a total of 175. No error message.

Now, if you login in a different browser with fresh cachet, the database upgrade screen turns up, but only for steps 5,6,7. If you click the upgrade button, nothing happens.

Missing tables are:
glbudgetdetails 
glbudgetheaders
gltotals

If you import these manually, a new login will show for 1 second the prior data and then vanish.

I do hope this feedback is useful.

Enrique

**Reply by timschofield** [reply]
Hello Enrique, thanks for your message.

Two questions to help me get started diagnosing this problem.

1 Have you managed to install a clean new database on this system?

2 What do the web server logs say when you are trying to login?

Thanks
Tim

**Reply by efeoli** [reply]
Hi Tim,

1. Yes, the installer runs well now and clean database with 178 tables is configured.
2. The server log file is attached for last attempt to import 4.15.2 database.

Enrique

[Nov-2024.tar.gz](https://github.com/user-attachments/files/17688868/Nov-2024.tar.gz)

**Reply by efeoli** [reply]
Hi Tim,
the update3 and 5 fix works for me. I will get back to you on Monday with the manufacturing issue and other minor issues.

Thank you so much for fixing the database updater.

Enrique

**Comment by dalers** [comment]
@timschofield - at the moment I can't login (username/password not accepted). It seems related to selecting the correct company and the sub-directories in companies/. It seems config.php no longer has an array of company names for selecting in the login screen. This is curious because I did login last night and the database update script did run (the exact same behavior as the beginning of this post).

Is there any special handing for companys that would be required when migrating from a 4.15.2 database (or 4.15 code) to the current system?

Once I get this sorted and I can login I'll send you the php errors.

Btw, it was convenient to be able to simply run the old SQL update scripts in the database monitor, but that can't be done now with the new PHP scripts (and that use library functions). Can you tell me the bash command line to run a new install script?

**Comment by timschofield** [comment]
The login screen checks for the sub-directories under the companies/ directory. These need to be the same name as the database. Optionally this sub-directory can contain a file called Companies.php. This should look something like the following:

```
<?php

$CompanyName['weberp'] = 'webERP Demonstration Company Ltd';

?>
```

where obviously weberp is the database name and webERP Demonstration Company Ltd is the full company name. If this exists, then the full company name is the one displayed in the login screen, otherwise what is displayed is simply the database name.

You cannot run the new php update files from outside webERP. The intention when designing this update system was that webERP would control it's own updating. The scripts also update the DBUdateNumber config option.

Thanks
Tim

ps Is there any chance you can let me have a copy of this DB so I can try it out myself?

**Comment by dalers** [comment]
>  ... file called Companies.php. This should look something like the following:

Did you forget to paste an example? ( I am now on my phone and don't see anything)

> any chance you can let me have a copy of this DB

Of course!  :-)  just look back a few replies to find an attached .sql.gz file

**Comment by timschofield** [comment]
Sorry Dale it didn't format the code properly, and just didn't show it. I've edited it and changed the formatting.

I must have missed that attachment, let me find it.

Tim

**Comment by timschofield** [comment]
OK, I took Dale's sql and did the following:

First I removed the USE weberp like in the sql file as I already have a db with that name - obviously the path names need changing

In the mysql command line:
CREATE DATABASE dale;
USE dale
\. ~/weberp_scc-20181001-weberp.sql  <- to load the sql
UPDATE www_users SET password='$2y$10$w87/ipUjZzNfDsWjNpLvuuG2MhwOtbt3mNroY0lI5/GSVxPGoy6JO' WHERE userid='admin'; <-To set the password to 'weberp'
\. ~/code/webERP/sql/mysql/upgrade4.15-4.15.1.sql  <- needed to bring the database up to 4.15.1 
\. ~/code/webERP/sql/mysql/upgrade4.15.1-4.15.2.sql <- needed to bring the database up to 4.15.2

Then from the Linux command line I ran:
mkdir companies/dale <- create a sub directory with the name I gave to the db
cp companies/weberp/* companies/dale -r <- copy all the contents of the weberp company to the dale directory

I changed the line in the companies/dale/Companies.php file to read:

$CompanyName['dale'] = 'Dales Company Ltd';

Then when I logged in with admin/weberp I was prompted to do the 7 database upgrades

Updates 3 & 5 had problems with strict mysql-mode, so I have pushed revised versions of these 2.

Having done this, I can now log back in.

**Reply by dalers** [reply]
Woo hoo! Thanks Tim. Will try later today (now at the first country Christmas market of the year)

**Comment by timschofield** [comment]
Enjoy the Christmas market.

I got half my Christmas shopping done today!!

**Comment by timschofield** [comment]
Glad it's working for you Enrique.

Many thanks for testing
Tim

On Sat, 9 Nov 2024 at 19:44, efeoli ***@***.***> wrote:

> Hi Tim,
> the update3 and 5 fix works for me. I will get back to you on Monday with
> the manufacturing issue and other minor issues.
>
> Thank you so much for fixing the database updater.
>
> Enrique
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/timschofield/webERP/discussions/210#discussioncomment-11199608>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAL6LH74EXOSFV36CX5VPDZ7ZQ2XAVCNFSM6AAAAABRMIFWUCVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMJZHE3DAOA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
www.weberp.org <http://www.weberpafrica.com>
@TimSchofield2
Blog: https://kwamoja.home.blog/

**Comment by dalers** [comment]
Hi Tim, found some nice olives but unfortunately no real Christmas shopping.

I sorted out the database and companies/ sub-directories and was able to login but got a blank screen again (HTTP ERROR 500).

The following error was in /var/log/httpd-error.log

`[Sat Nov 09 18:54:13.641103 2024] [php:error] [pid 85772] [client 174.0.43.39:62983] PHP Fatal error:  Uncaught mysqli_sql_exception: Unknown column 'commissionsact' in 'field list' in /usr/local/www/weberp/includes/ConnectDB_mysqli.inc:58\nStack trace:\n#0 /usr/local/www/weberp/includes/ConnectDB_mysqli.inc(58): mysqli_query()\n#1 /usr/local/www/weberp/includes/GetConfig.php(84): DB_query()\n#2 /usr/local/www/weberp/includes/UserLogin.php(169): include('...')\n#3 /usr/local/www/weberp/includes/session.php(171): userLogin()\n#4 /usr/local/www/weberp/index.php(4): include('...')\n#5 {main}\n  thrown in /usr/local/www/weberp/includes/ConnectDB_mysqli.inc on line 58, referer: https://weberp.dalescott.net/`

I reloaded the page which resulted in Z_UpgradeDatabase.php. Continue with Updates appears to have completed successfully so your changes to the update scripts worked on my system also.

![image](https://github.com/user-attachments/assets/65fe8176-0492-4f30-bacf-a1fe76e1c755)

I clicked "You need to logout..." and logged back in, and this time the main menu appeared instead of HTTP ERROR 500.

However, there was also a new error in /var/log/httpd-error.log:

`[Sat Nov 09 19:07:32.754995 2024] [php:error] [pid 88551] [client 174.0.43.39:63081] PHP Fatal error:  Uncaught mysqli_sql_exception: Table 'weberp_scc.modules' doesn't exist in /usr/local/www/weberp/includes/ConnectDB_mysqli.inc:58\nStack trace:\n#0 /usr/local/www/weberp/includes/ConnectDB_mysqli.inc(58): mysqli_query()\n#1 /usr/local/www/weberp/includes/MainMenuLinksArray.php(13): DB_query()\n#2 /usr/local/www/weberp/index.php(7): include('...')\n#3 {main}\n  thrown in /usr/local/www/weberp/includes/ConnectDB_mysqli.inc on line 58, referer: https://weberp.dalescott.net/`

Casually poking around the menus didn't cause any more errors (either on the screen or in httpd-errors.log).

Any ideas what caused the two initial errors?

Cheers,
Dale

**Comment by timschofield** [comment, marked as answer]
The commissionsact account error was shown because that field wasn't created until 1.php was run. So I have put in includes/GetConfig.php a check to not load it unless 1.php was already run.

The modules table not existing message was created because the Z_UpdateDatabase.php script was trying to load them. I have forced the script to exit before it gets this far. This change is in includes/session.php.

Thanks
Tim

**Comment by dalers** [comment]
Thanks Tim. I pulled the latest commits and successfully repeated the migration with no errors or hiccups! :-)

**Comment by timschofield** [comment]
Good stuff Dale. Hopefully you can test the system now.

Tim

On Mon, 11 Nov 2024 at 03:57, Dale Scott ***@***.***> wrote:

> Thanks Tim. I pulled the latest commits and successfully repeated the
> migration with no errors or hiccups! :-)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/timschofield/webERP/discussions/210#discussioncomment-11210158>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAL6LBVL4L3LGBYK3VRRFD2AATJDAVCNFSM6AAAAABRMIFWUCVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMRRGAYTKOA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
www.weberp.org <http://www.weberpafrica.com>
@TimSchofield2
Blog: https://kwamoja.home.blog/

**Comment by dalers** [comment]
But Tim, I thought I already was testing... <wink>

I'm going to start repeating my old webERP tutorial blog posts to test basic purchasing and production and to update the posts with new 
screenshots.

After that, I'd like to write two new posts for testing selling and another to test basic financial reporting including closing a period.

Do you have any preferences for what to test after that?

Cheers,
Dale

**Comment by timschofield** [comment]
Hi Dale, I meant now that I had finally got the installer working you
can start with the system proper!

I think a great way to test is a complete run through of a particular
scenario. So for instance, create a part, create a supplier, raise a
purchase order for that part, from that supplier, receive the purchase
order into stock, create a customer, create a sales order of this part
to this customer, despatch the part, invoice the customer. Checking
stuff like authorisation along the way.

Thanks,
Tim



On Tue, 12 Nov 2024 at 03:29, Dale Scott ***@***.***> wrote:
>
> But Tim, I thought I already was testing... <wink>
>
> I'm going to start repeating my old webERP tutorial blog posts to test
> basic purchasing and production and to update the posts with new
> screenshots.
>
> After that, I'd like to write two new posts, one on selling and another
> on basic financial reporting including closing a period.
>
> Do you have any recommendations for what to test next?
>
> Cheers,
> Dale
> On 2024-11-11 11:07, Tim Schofield wrote:
>
> > Good stuff Dale. Hopefully you can test the system now.
> >
> > Tim
> >
> > On Mon, 11 Nov 2024 at 03:57, Dale Scott ***@***.***> wrote:
> >
> >> Thanks Tim. I pulled the latest commits and successfully repeated the
> >> migration with no errors or hiccups! :-)
> >>
> >> --
> >> Reply to this email directly, view it on GitHub
> >> <https://github.com/timschofield/webERP/discussions/210#discussioncomment-11210158>,
> >> or unsubscribe
> >> <https://github.com/notifications/unsubscribe-auth/AAAL6LBVL4L3LGBYK3VRRFD2AATJDAVCNFSM6AAAAABRMIFWUCVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMRRGAYTKOA>
> >> .
> >> You are receiving this because you were mentioned.Message ID:
> >> ***@***.***>
> >>
> >
> > --
> > www.weberp.org <http://www.weberpafrica.com>
> > @TimSchofield2
> > Blog: https://kwamoja.home.blog/
> > --
> > Reply to this email directly, view it on GitHub [1], or unsubscribe
> > [2].
> > You are receiving this because you authored the thread.Message ID:
> > ***@***.***>
>
>
> Links:
> ------
> [1]
> https://github.com/timschofield/webERP/discussions/210#discussioncomment-11217612
> [2]
> https://github.com/notifications/unsubscribe-auth/AAKFYYD3O4QWQTYNU2GSI2L2ADW4ZAVCNFSM6AAAAABRMIFWUCVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMRRG43DCMQ
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.Message ID: ***@***.***>



-- 
www.weberp.org
@TimSchofield2
Blog: https://kwamoja.home.blog/
